include/pybind11/detail/type_caster_base.h

```diff
@@ -739,16 +739,9 @@ class type_caster_generic {
     // isn't needed or can't be used.  If the type is unknown, sets the error and returns a pair
     // with .second = nullptr.  (p.first = nullptr is not an error: it becomes None).
     PYBIND11_NOINLINE static std::pair<const void *, const type_info *> src_and_type(
-            const void *src, const std::type_info &cast_type, const std::type_info *rtti_type = nullptr) {
-        if (auto *tpi = get_type_info(cast_type))
-            return {src, const_cast<const type_info *>(tpi)};
-
-        // Not found, set error:
-        std::string tname = rtti_type ? rtti_type->name() : cast_type.name();
-        detail::clean_type_id(tname);
-        std::string msg = "Unregistered type : " + tname;
-        PyErr_SetString(PyExc_TypeError, msg.c_str());
-        return {nullptr, nullptr};
+            const void *src, const std::type_info &cast_type) {
+        auto *tpi = get_type_info(cast_type, true);
+        return {src, const_cast<const type_info *>(tpi)};
```

Comment:
I just noticed that this change should be applied to master. I will file a separate PR for that.